Testing module to test the person extraction

BenjaminBlaesi opened the pull request:
## What?

This is a helper module that allows to test the implementation of the `court_composition_extracting_functions` and the `procedural_participation_extracting_functions` .

## Why?

It's useful for debugging and testing. It can be used to check if a new implementation gives the expected output, and to check if changes to an existing implementation don't break the expected result. 

## How?

For a specific spider, an example section taken from the section splitting is given to the corresponding extracting function, and with multiple assertions it is checked if the result is correct. If it isn't, an error is displayed. 

## Anything Else?

The tests I wrote fail because they were written for a different PR that isn't merged yet and that I have to clean up first.

BenjaminBlaesi commented:
@Skatinger  : I put my testing functions into a separate module as you suggested. Let me know what you think of it. 
@susifohn : You were also interested in the testing functions, I think.

Skatinger commented:
Great, I love it!

##### Things I suggest should be changed before merge:
- [ ] I saw you added a small doc with the functions, that's already great! One thing that usually helps a lot for users is a short explanation on how and when to use your functions. For example how do I call them? can I import them and write them inside my code? Or can I call them with python -m path-to-test-module <test-name>.  I think in your case you should also mention how one can add a new test case. (I saw the "uncomment this" at the invocations, but they should probably be mentioned at the top, so users don't have to scroll all the way through, or can access your docs via the help(yourmodule) command [see here](https://docs.python.org/3/library/functions.html#help).
- [ ] allow to run the tests without commenting / uncommenting (see [python modules](https://docs.python.org/3/tutorial/modules.html#executing-modules-as-scripts)) This will allow to run tests programmatically instead of invoking them manually with code changes. Commenting/Uncommenting is bad practice.
- [ ] Maybe you can manage to do your assertions with [unit tests](https://docs.python.org/3/library/unittest.html). In your implementation I'm afraid the test will fail upon the first failing assertion, but when testing we usually want a summary of all failed and suceeded tests, so we can address them all.

##### A few remarks (can also be addressed later):
- The file test_utils.py will probably not be the best place for all tests. As testing will increase and tests for other modules will be implemented a clean folder structure would be more suitable. for example `root/test/preprocessors/judicial_person_extraction_test.py`
- make the tests dynamic, so that one can pass arguments instead of adding code to the test function itself.

BenjaminBlaesi commented:
Thank you for your helpful and detailed suggestions, @Skatinger

Changes:

- I improve the documentation and added how to use them and how to add new tests. Did you want to import the tests and use them in your code?
- The tests can now be run all together or separately.
- The tests now use unittest.

Remarks:

- Possibly. I don't know what kind of directory structure @JoelNiklaus  prefers. 
- What kind of arguments do you want to pass?

JoelNiklaus commented:
Great collaboration :)

There are two folder structures for the tests (https://docs.pytest.org/en/6.2.x/goodpractices.html): 
1. Tests outside application code
2. Tests as part of application code

I personally like to have tests separate (so that would be option 1). But since, I am by no means an expert in tests, I would probably be very easily persuaded if you motivate option 2 ;)

Skatinger commented:
> What kind of arguments do you want to pass?

You defined your input at the top of the file, which in my eyes is not perfect but ok. Usually you would generate your test case data right within a test or a specific test context. It's also possible to do it over a whole module like you do know, so not too much of a problem. Maybe look at setup and teardown for tests.

My concern is with the ground-truth. Instead of writing a test function for every single court, you could write a single one which takes several arguments. That way you can remove a lot of duplication. Snippet to show what I mean:

```python
class MyTestModule(unittest.TestCase):
  # setup function to create all the  necessary data for our tests
  def setUp(self):
    # define the test data
    test_data = { zh_steuerrekurs: "bli bla blu, some string", ... }
    # define expected result for each test
    expected_results = { zh_steuerrekurs: { judges: ["bli", "bla"], clerks: ["blu"], other: "some string" } }, ... }

  # define a test function within testing module to check all the given data
  def test_person_extraction(self, spider_name):
      # now we can test any passed spider
      result = extraction.locals()[spider](test_data[spider])
      assertEquals(result, expected_results[spider]

# then in main you can call it for every available spider
for spider in test_data.keys():
  test_person_extraction(spider)
```

Does not have to be like this, but as you can see it generally makes things easier for use later on. Instead of adding a whole new method for my test-case, i can just add an entry to the data dictionaries.

One thing I should probably mention is that the assertEquals probably will not work too well, but you can always write several methods, like checking the intersection between roles you found and roles that should be present, or check if the found names are correct etc. I would start with some simple cases, which can be detailed later on.

As I already said, I think in it's current form its not bad at all. It's well documented, it works and is structured. That is already the most important thing. The rest is good practice which will make it better in the future, but for now the benefit is not too big I guess.

Thanks for addressing all the feedback by the way, looks much better already! Will probably use it in January for my spiders if I get to it.

@BenjaminBlaesi

susifohn commented:
@BenjaminBlaesi  Yes, very good to have unit testing. Seems very promising.

About the question for having the test code our/inside source code, I highly recommend to have it separated from the source code (so called application code). It's also good practice to distinguish between unit and module testing. And of course, our unit tests have to runn fast. :-)

BenjaminBlaesi commented:
@Skatinger 
Thank you for your extensive comment. 
I've now converted the test module from unittest to pytest. pytest seems to be a bit more common as a standard.
I've rewritten the tests so that new tests can be written without adding new functions. For now, every test only tests for equality, anyway. 
But I wanted to be able to test specific values of the result and not just compare the entire result. This allows more flexibility in testing and also returns more concrete error messages when testing. 
You still need to add a test header and one line to setup the test, but the rest is just added to the test_data list, where you can specify the key of the object you want to test and the value you expect as a result.
This should make it easier to add new tests and makes the tests more compact, but the tests are still flexible and it's easy to see which results you are expecting.

BenjaminBlaesi commented:
I also had to add an empty conftest.py file so that pytest could handle the paths of the imports. And I changed the file name to test_utils because pytest wants the file name to start with test_ and not with testing_.

JoelNiklaus commented on scrc/utils/test_utils.py:
```diff
@@ -0,0 +1,208 @@
+import scrc.preprocessors.extractors.spider_specific.procedural_participation_extracting_functions as p
+
+"""
+    Helper module to test the functions of other modules.
+    Allows to check if the court composition extraction and the procedural participation extraction returns the expected result for a given test input.
+
+    Classes:
+        `TestCourtCompositionExtractingFunctions`
+        `TestProceduralParticipationExtractingFunctions`
+
+    Usage:
+        Assuming you are in the SwissCourtRulingCorpus directory and use the command line:
+        Run all tests with `pytest`
+        Run only this module with `pytest scrc/utils/test_utils.py`
+        Run a specific class with `pytest -k "ClassName"`
+            e.g. `pytest -k "TestCourtCompositionExtractingFunctions"`
+        Run a specific function with `pytest -k "ClassName and function_name"`
+            e.g. `pytest -k "TestCourtCompositionExtractingFunctions and test_equality"`
+        Consult `https://docs.pytest.org/en/6.2.x/usage.html` for more information.
+
+    Adding tests:
+        Add tests for a new module by creating a new class.
+        Add tests for a spider by 
+            - adding new test headers
+            - adding a new setup line in the module that you want to test
+            - adding what you want to test and what you expect to the test_data
+        Add new kinds of test by adding new functions.
```
Great comment!

BenjaminBlaesi replied:
Thank you!

JoelNiklaus commented on scrc/utils/test_utils.py:
```diff
@@ -0,0 +1,208 @@
+        # ZG_Verwaltungsgericht
+        (zg_vg.president.name, 'Adrian Willimann'),
+        (zg_vg.judges[0].name, 'Adrian Willimann'), 
+        # (zg_vg.judges[1].name, 'Jacqueline Iten-Staub'),
+        # (zg_vg.judges[2].name, 'Matthias Suter'),
+        # (zg_vg.clerks[0].name, 'Patrick Trütsch'),
```
Are there reasons for these commented lines?

BenjaminBlaesi replied:
Yes, these lines caused some index out of bounds errors during testing, so I commented them to see if the other tests run (which they do). The tests should all work after merging due to the changes I made in other PRs; and I meant to uncomment them again which I now did. 

Or should I have first merged those changes into this branch? I thought it was simpler not to do so if this branch has no conflicts with the base branch.

JoelNiklaus replied:
I see. Makes sense, it's fine that way.

JoelNiklaus commented on scrc/utils/test_utils.py:
```diff
@@ -0,0 +1,208 @@
+        (zh_svg.president.gender.value, 'male'),
+        (zh_svg.judges[0].name, 'Mosimann'),
+        (zh_svg.judges[0].gender.value, 'male'),
+        # (zh_svg.judges[1].name, 'Käch'),
+        # (zh_svg.judges[1].gender.value, 'female'),
+        # (zh_svg.judges[2].name, 'Sager'),
+        # (zh_svg.judges[2].gender.value, 'female'),
+        # (zh_svg.clerks[0].name, 'Kudelski'),
+        # (zh_svg.clerks[0].gender.value, 'female'),
```
Same here. Maybe add a short comment explaining why they are commented. If they are not relevant, wrong or similar, consider removing them

BenjaminBlaesi replied:
Same as above.

JoelNiklaus reviewed (commented):
Looks great. I have just a quick question regarding the commented lines (see comments in code).

JoelNiklaus commented on scrc/utils/test_utils.py:
```diff
@@ -82,12 +82,14 @@ def court_composition_setup(court, test_list, namespace):
     zh_svg = court_composition_setup(c.ZH_Sozialversicherungsgericht, ZH_Sozialversicherungsgericht_test_header, namespace_de)
 
     test_data = [
+        # The first element of each tuple is the value to be tested,
+        # and the second element is the expected result.
```
Thanks for that explanatory comment
